packages/react/src/helpers/utils.ts

```diff
@@ -1,8 +1,30 @@
+import isEmpty from 'lodash/isEmpty';
+import isArray from 'lodash/isArray';
+import isObject from 'lodash/isObject';
+
 export const isDevelopment = () => process.env.NODE_ENV !== 'production';
 
+const isEmptyObj = (val: any) => isObject(val) && isEmpty(val);
+
+const isEmptyArr = (val: any) => isArray(val) && isEmpty(val);
+
 export const areArraysEqual = (arr1: Array<any>, arr2: Array<any>) => {
   if (arr1.length !== arr2.length) return false;
-  return arr1.every((value, index) => value === arr2[index]);
+  return arr1.every((elem1, index) => {
+    const elem2 = arr2[index];
+    /**
+     * edge cases: if both values are empty object/array, we consider them equal.
+     * These can catch empty default values (`[]`, `{}`) that unintentionally point
+     * to different refernces.
+     *
+     * We can consider doing a deep comparison, but left it here for efficiency
+     * + practicality for authenticator state comparison purposes.
+     */
```

Comment:
This optimization is particularly targeted against comparing default empty values, like https://github.com/aws-amplify/amplify-ui/blob/e20720a894ccb2cfcc7ede7160299e082ec76fd2/packages/ui/src/helpers/authenticator/facade.ts#L59

which will always create a **new** empty object even if `validationError` doesn't exist, and thus would fail shallow equality check. This would cause a re-render every single state change! Adding these empty arr/obj checks will optimize these cases.

Reply:
Alternatively, we could do a deep value comparison with something like `lodash/isEqual`. But this is expensive for large objects like `user`, so I left it at shallow comparison + treating empty array/object as equal.

Reply:
👍